src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -314,6 +342,49 @@ def classpath_dest_filename(coord: str, src_filename: str) -> str:
     return f"{dest_name}{ext}"
 
 
+@dataclass(frozen=True)
+class ArtifactRequirementsWithLocalFiles:
+    # omg tidy up this name. it's bad
+    artifact_requirements: ArtifactRequirements
+    digest: Digest
+
+
+@rule
+async def use_local_artifacts_where_possible(
+    input_requirements: ArtifactRequirements,
+    all_jars: AllJarTargets,
+) -> ArtifactRequirementsWithLocalFiles:
+    output: List[Coordinate] = []
+    further_processing: List[Target] = []
+
+    for req in input_requirements:
+        tgt = all_jars.by_coordinate.get(req)
+        if not tgt:
+            output.append(req)
+        else:
+            further_processing.append(tgt)
+
+    files = await Get(
+        SourceFiles,
+        SourceFilesRequest(tgt[JvmArtifactJarSourceField] for tgt in further_processing),
+    )
+
+    for target, file in zip(further_processing, files.files):
+        coord = Coordinate.from_jvm_artifact_target(target)
+        coord = Coordinate(
```

Comment:
Use `dataclasses.replace`. Note: by going field by field, this code missed the `Coordinate.packaging` field.